src/main/java/edu/ucla/library/iiif/auth/services/DatabaseServiceImpl.java

```diff
@@ -0,0 +1,239 @@
+import io.vertx.core.json.JsonObject;
+import io.vertx.pgclient.PgConnectOptions;
+import io.vertx.pgclient.PgPool;
+import io.vertx.redis.client.Redis;
+import io.vertx.redis.client.RedisOptions;
+import io.vertx.sqlclient.PoolOptions;
+import io.vertx.sqlclient.Row;
+import io.vertx.sqlclient.RowSet;
+import io.vertx.sqlclient.Tuple;
+
+/**
+ * The implementation of DatabaseService.
+ */
+public class DatabaseServiceImpl implements DatabaseService {
+
+    /**
+     * The database service's logger.
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(DatabaseService.class, MessageCodes.BUNDLE);
+
+    /**
+     * The postgres database (and default user) name.
+     */
+    private static final String POSTGRES = "postgres";
+
+    /**
+     * The name of the database's "items" table.
+     */
+    private static final String ITEMS = "items";
+
+    /**
+     * The name of the database's "origins" table.
+     */
+    private static final String ORIGINS = "origins";
+
+    /**
+     * The PreparedQuery template for selecting an item's "access level".
+     */
+    private static final String SELECT_ACCESS_LEVEL = "SELECT access_level FROM items WHERE uid = $1";
+
+    /**
+     * The PreparedQuery template for upserting an item's "access level".
+     */
+    private static final String UPSERT_ACCESS_LEVEL = String.join(SPACE, "INSERT INTO items VALUES ($1, $2)",
+            "ON CONFLICT (uid) DO", "UPDATE SET access_level = EXCLUDED.access_level");
+
+    /**
+     * The PreparedQuery template for selecting an origin's "degraded allowed".
+     */
+    private static final String SELECT_DEGRADED_ALLOWED = "SELECT degraded_allowed FROM origins WHERE url = $1";
+
+    /**
+     * The PreparedQuery template for upserting an origin's "degraded allowed".
+     */
+    private static final String UPSERT_DEGRADED_ALLOWED = String.join(SPACE, "INSERT INTO origins VALUES ($1, $2)",
+            "ON CONFLICT (url) DO", "UPDATE SET degraded_allowed = EXCLUDED.degraded_allowed");
+
+    /**
+     * The database's default hostname.
+     */
+    private static final String DEFAULT_HOSTNAME = "localhost";
+
+    /**
+     * The underlying SQL client.
+     */
+    private final PgPool myDbConnectionPool;
+
+    /**
+     * The underlying Redis client.
+     */
+    private final Redis myDbCacheClient;
+
+    /**
+     * Creates an instance of the service.
+     *
+     * @param aVertx A Vert.x instance
+     * @param aConfig A configuration
+     */
+    public DatabaseServiceImpl(final Vertx aVertx, final JsonObject aConfig) {
+        myDbConnectionPool = PgPool.pool(aVertx, getConnectionOpts(aConfig), getPoolOpts(aConfig));
+        myDbCacheClient = Redis.createClient(aVertx, getDbCacheClientOpts(aConfig));
+    }
+
+    @Override
+    public Future<Void> close() {
+        myDbCacheClient.close();
+        return myDbConnectionPool.close();
+    }
+
+    @Override
+    public Future<Integer> getAccessLevel(final String aId) {
```

Comment:
I don't know that we have this documented anywhere, but I've been going with capitalized forms of variables at the end of a camel-case variable so, for instance, `aIdOfGreatImportance` and `aID`. Wondering how you feel about this or whether it's better to just be consistent with the camel-case regardless of location in the variable name(?)

Reply:
Ah right, we did discuss this before. I'm not too opinionated about this so am happy to go along with your naming convention.